### PR TITLE
Change assert(is_err()) to assert_noop to check state consistency on errors

### LINKED ISSUE #8545: Audit use of `.is_err()` in Runtime Tests
When testing extrinsics, it is important that we follow the "Check First, Write Last" approach so that an extrinsic which results in an error does NOT modify the runtime state.

There is a special macro which does this: `assert_noop!`.

This function specifically checks that a function results in error and that the state root of the blockchain before and after that error is the same.

There may be some places in runtime tests where `asssert!(function.is_err())` is used instead of `assert_noop!`. This can hide such problems about modifying the state, and all of these instances should be replaced with `assert_noop!` instead.

### PR DESCRIPTION
**Pull request by hirschenberger**
fixes #8545

### CODE REVIEW

File: frame/transaction-payment/src/lib.rs [hunk 1]
```diff
@@ -616,6 +616,7 @@ mod tests {
 	use sp_runtime::{
 		testing::{Header, TestXt},
 		traits::{BlakeTwo256, IdentityLookup},
+        transaction_validity::InvalidTransaction,
```
Comment (kianenigma) [inline frame/transaction-payment/src/lib.rs:619]:
```suggestion
		transaction_validity::InvalidTransaction,
```

**Review by kianenigma — APPROVED** [review]

**Comment by hirschenberger** [comment]
That failing test is strange, it sometimes works and sometimes not. Maybe due to sccache?

**Comment by hirschenberger** [comment]
I think I need some help here with the failing test.

When I change the expected error type to a wrong value e.g. `TransactionValidityError::Invalid(InvalidTransaction::BadProof)` I get this, which is what I expect:

```
---- tests::bad_extrinsic_not_inserted stdout ----
on_initialize(1)
thread 'tests::bad_extrinsic_not_inserted' panicked at 'assertion failed: `(left == right)`
  left: `Err(TransactionValidityError::Invalid(InvalidTransaction::Future))`,
 right: `Err(TransactionValidityError::Invalid(InvalidTransaction::BadProof))`', frame/executive/src/lib.rs:860:13
 ```
 
 When I change the type to the right value: `TransactionValidityError::Invalid(InvalidTransaction::Future)` I get this strange error:
 
 ```
 ---- tests::bad_extrinsic_not_inserted stdout ----
on_initialize(1)
thread 'tests::bad_extrinsic_not_inserted' panicked at 'assertion failed: `(left == right)`
  left: `[43, 22, 252, 42, 247, 106, 220, 153, 188, 63, 218, 120, 117, 228, 9, 45, 42, 26, 239, 243, 16, 226, 217, 97, 143, 195, 251, 246, 181, 70, 68, 178]`,
 right: `[65, 171, 6, 54, 28, 188, 146, 219, 140, 67, 51, 124, 41, 234, 137, 189, 241, 58, 79, 94, 30, 245, 108, 142, 60, 124, 58, 24, 198, 160, 189, 158]`', frame/executive/src/lib.rs:860:13
 ```

**Comment by shawntabrizi** [comment]
@hirschenberger this means you caught a bug :)

**Comment by hirschenberger** [comment]
Yippie ;-)

Any hints where it goes wrong? The lists are no UTF-8 strings and I assume also no SCALE encoded data. Do you think it's a substrate or rust issue?

**Comment by shawntabrizi** [comment]
@hirschenberger it means that in the test, there is some extrinic which is being called, and returning an error. But it is also changing the state of the chain. This is ultimately what the `noop` assert tests for.

You need to look into the extrinsic to see if you can catch where it can fail after already changing state. I will look into this now, it can be a little complex and sensitive to solve these issues.

BTW can you please include your DOT / KSM address? Have you done that already in another PR?

**Comment by hirschenberger** [comment]
I see, excited to see what the reason for the state change is. I'll also look in the code and try to help with the analysis.

Yes I already got a tip from you in another PR: https://github.com/paritytech/substrate/pull/8541#issuecomment-815245007

**Comment by gavofyork** [comment]
There is no bug here.

The function in question (`apple_extrinsic`) is not a dispatchable, therefore it does not conform to the usual no-change-on-error. `apply_extrinsic` is allowed to, and indeed *will*, mutate the state if given an erroring transaction since it is assumed that the transaction is added to the block regardless of whether it happened to be successful or not.

File: frame/contracts/src/exec.rs [hunk 2]
```diff
@@ -1571,7 +1572,8 @@ mod tests {
 			);
 
 			// Check that the account has not been created.
-			assert!(Storage::<Test>::code_hash(&instantiated_contract_address).is_err());
+			assert_noop!(Storage::<Test>::code_hash(&instantiated_contract_address),
+				ContractAbsentError);
```
Comment (shawntabrizi) [inline frame/contracts/src/exec.rs:1576]:
```suggestion
			assert_noop!(
				Storage::<Test>::code_hash(&instantiated_contract_address),
				ContractAbsentError,
			);
```

or put it on a single line

File: frame/democracy/src/benchmarking.rs [hunk 3]
```diff
@@ -206,11 +207,12 @@ benchmarks! {
...
 		let referendum_index = add_referendum::<T>(0)?;
 		let call = Call::<T>::emergency_cancel(referendum_index);
-		assert!(Democracy::<T>::referendum_status(referendum_index).is_ok());
+		assert_ok!(Democracy::<T>::referendum_status(referendum_index));
 	}: { call.dispatch_bypass_filter(origin)? }
 	verify {
 		// Referendum has been canceled
-		assert!(Democracy::<T>::referendum_status(referendum_index).is_err());
+		assert_noop!(Democracy::<T>::referendum_status(referendum_index),
+            Error::<T>::ReferendumInvalid);
```
Comment (shawntabrizi) [inline frame/democracy/src/benchmarking.rs:215]:
```suggestion
		assert_noop!(
			Democracy::<T>::referendum_status(referendum_index),
			Error::<T>::ReferendumInvalid,
		);
```

Comment (shawntabrizi) [reply frame/democracy/src/benchmarking.rs:215]: Probably other such formatting things, wont look at them all.

Comment (hirschenberger) [reply frame/democracy/src/benchmarking.rs:215]: I thought this is the coding style, I can go over the changes and fix it everywhere.

File: frame/transaction-payment/src/lib.rs [hunk 4]
```diff
@@ -927,10 +923,10 @@ mod tests {
 				class: DispatchClass::Normal,
 				pays_fee: Pays::Yes,
 			};
-			assert!(
+			assert_noop!(
 				ChargeTransactionPayment::<Runtime>::from(0)
-					.validate(&1, CALL, &free_transaction , len)
-					.is_err()
+					.validate(&1, CALL, &free_transaction , len),
+				TransactionValidityError::Invalid(InvalidTransaction::Payment)
```
Comment (shawntabrizi) [inline frame/transaction-payment/src/lib.rs:929]:
```suggestion
				TransactionValidityError::Invalid(InvalidTransaction::Payment),
```

**Comment by shawntabrizi** [comment]
@hirschenberger no force push please :)

If you need help with git, happy to do so.

**Comment by hirschenberger** [comment]
> @hirschenberger no force push please :)
> 
> If you need help with git, happy to do so.
Ok, no force pushing and no squashing, so it's ok to accumulate several commits in a PR?

**Comment by hirschenberger** [comment]
I think I'm done. CI got stuck.

**Review by shawntabrizi — APPROVED** [review]

**Comment by thiolliere** [comment]
bot merge

**Pull request merged**